src/VisualStudio/Core/Def/DesignerAttribute/VisualStudioDesignerAttributeService.cs

```diff
@@ -111,79 +133,180 @@ private async ValueTask ProcessWorkspaceChangeAsync(CancellationToken cancellati
+            // semantic-version of any dependent projects change.  The reason for checking dependent projects is that we
+            // look for the designer attribute on subclasses as well (so we have to walk the inheritance tree).  This
+            // tree may be unfortunately be affected by dependent projects.  In an ideal design we would require the
+            // attribute be on the declaration point so we could only check things that are known to directly have an
+            // attribute on them, and we wouldn't have to look up the inheritance hierarchy.
+            var dependentSemanticVersion = await project.GetDependentSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+
+            if (!_projectToLastComputedDependentSemanticVersion.TryGetValue(project.Id, out var lastComputedVersion) ||
+                lastComputedVersion != dependentSemanticVersion)
+            {
+                var stream = client.TryInvokeStreamAsync<IRemoteDesignerAttributeDiscoveryService, DesignerAttributeData>(
+                    project,
+                    (service, checksum, cancellationToken) => service.DiscoverDesignerAttributesAsync(checksum, project.Id, priorityDocumentId, cancellationToken),
+                    cancellationToken);
+
+                await ProcessResultsDataAsync(project.Id, stream).ConfigureAwait(false);
+
+                // now that we're done processing the project, record this version-stamp so we don't have to process it again in the future.
+                _projectToLastComputedDependentSemanticVersion[project.Id] = dependentSemanticVersion;
+            }
+        }
+
+        private async Task ProcessResultsDataAsync(ProjectId projectId, IAsyncEnumerable<DesignerAttributeData> stream)
+        {
+            using var _ = PooledHashSet<DocumentId>.GetInstance(out var seenDocuments);
 
-            await client.TryInvokeAsync<IRemoteDesignerAttributeDiscoveryService>(
-                solution,
-                (service, checksum, callbackId, cancellationToken) => service.DiscoverDesignerAttributesAsync(callbackId, checksum, priorityDocument, cancellationToken),
-                callbackTarget: this,
-                cancellationToken).ConfigureAwait(false);
+            // get the results and add all the documents we hear about to the notification queue so they can be batched up.
+            await foreach (var data in stream.ConfigureAwait(false))
+            {
+                seenDocuments.Add(data.DocumentId);
+                _projectSystemNotificationQueue.AddWork(data);
+            }
+
+            // now, if we didn't hear about some document that we previous reported, then go through and notify the project system
+            // that the document is no longer designable.
+            foreach (var (documentId, lastReportedData) in _documentToLastReportedData)
+            {
+                if (documentId.ProjectId == projectId &&
+                    !seenDocuments.Contains(documentId))
```

Comment:
should we check that the document still exists before reporting a null category, or does the project system handle that

Reply:
we check for the legacy project system (where we're on the UI thread, and where we need an ItemId).  But we don't check on hte CPS case because it's inherently 'racey'.  we're issueing on the BG, so there's no way for us to ever know for sure.  the API is resilient to this though.